include/aero/grid/grid.hpp

```diff
@@ -1,44 +1,54 @@
 #ifndef AERO_GRID_HPP
 #define AERO_GRID_HPP
 
+#include <aero/array/array.hpp>
+
+#include <memory>
+
 namespace aero {
 
-class Grid {
+/// This type represents a computional grid consisting of a set of contiguous
+/// segments separated by interfaces. The grid keeps track of the midpoints
+/// between interfaces.
+struct Grid final {
 public:
-  /// Default constructor.
-  Grid() = default;
-  /// Copy constructor
-  Grid(const Grid& other) = default;
+  // A grid may be neither default constructed nor copied from another grid.
+  Grid() = delete;
+  Grid(const Grid&) = delete;
+  Grid& operator=(const Grid&) = delete;
+
+  /// Creates a grid from an Array containing points delimiting the interfaces
+  /// between segments. The Grid assumes ownership of this array.
+  explicit Grid(Array* interfaces);
+
+  /// Creates a grid from a C pointer. The grid data is copied so it lives in
+  /// both C and C++ runtimes.
+  static Grid from_c_ptr(void *c_ptr);
+
+  /// Creates a grid from a Fortran pointer. The Grid data is copied so it
+  /// lives in both Fortran and C++ runtimes.
+  static Grid from_fortran_ptr(void *f_ptr);
+
   /// Move constructor
   Grid(Grid&& other) = default;
 
   /// Destructor
-  virtual ~Grid() {}
+  ~Grid();
 
-  /// Assignment operator
-  Grid& operator=(const Grid&) = default;
   /// Default move assignment operator
   Grid& operator=(Grid&&) = default;
-};
-
-/// This Grid subclass allows access to arrays implemented in Fortran.
-class FortranGrid: public Grid {
-public:
-
-  /// Construct an aerosol model wrapped around a Fortran implementation
-  /// that can be accessed with the given pointer.
-  explicit FortranGrid(void *fortran_pointer);
-
-  // Overridden functionality
 
-  FortranGrid(FortranGrid& other);
-  FortranGrid(FortranGrid&& other);
-  ~FortranGrid() override;
+  /// Array storing interface coordinates (in ascending order)
+  std::unique_ptr<Array> interfaces;
 
-  FortranGrid& operator=(FortranGrid& other);
-  FortranGrid& operator=(FortranGrid&& other);
+  /// Array storing coordinates of midpoints between interfaces (in ascending
+  /// order)
+  std::unique_ptr<Array> midpoints;
 
-  void *f_ptr_; // pointer to Fortran grid implementation
+  /// Lower bound (minimum interface coordinate), provided for convenience
+  Real lower_bound;
+  /// Upper bound (maximum interface coordinate), provided for convenience
+  Real upper_bound;
```

Comment:
I tried making them `const` at first, but `const`ness has a lot of warts, and actually started generating a lot more noise than I wanted, so I backed off. We could try again, though--if we adopted a zero-copy `Grid` container pattern we might be able to do it.

Reply:
This kind of thing also seems fairly important for this audience to me. This is a group that has a long history of doing whatever it takes to get something working to publish a paper, possibly with the intent of fixing it, but the hack finding its way into production code where it later causes problems. e.g., I need to have this thing calculated assuming the width of the 5th grid segment is smaller to fit the data, and then five years later someone continuing to use the same object not realizing this change has been made.

Reply:
Yeah, it's definitely worth striving for. Most of the problems with `const` members in C++ classes arise during initialization. An easy way to keep these members const would be just to make them private and expose them only in `const` accessor functions instead of giving direct access to them. It's a bit old-fashioned in the C++ world, but maybe it's worth doing here, since interoperability makes it harder to do everything in a modern way.